### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML validation regex**

The title validation uses `/<[^>]*>/g` which correctly catches all HTML tags like ``, but the description validation uses `/<.*>/g` which has different behavior. This inconsistency could lead to security issues where HTML tags are blocked in titles but allowed in descriptions.

**Recommendation:** Use the same regex pattern for both fields:
```typescript
} else if (/<[^>]*>/g.test(description)) {
```

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Bug: ID "0" validation issue**

The current regex `/^\d+$/` allows "0" as a valid ID, but depending on your ID generation strategy (auto-increment starting from 1), ID "0" might not be valid. Consider if you need to explicitly exclude zero:

```typescript
if (!id || !/^[1-9]\d*$/.test(id)) {
```

Or add a separate check:
```typescript
const idNum = parseInt(id, 10);
if (!id || !/^\d+$/.test(id) || idNum === 0) {
```

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Bug: Missing error handling**

The `todoStore.create()` call is not wrapped in a try-catch block. If the store throws an exception, it could crash the server instead of returning a proper error response.

**Recommendation:**
```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({
    error: 'Internal server error',
    details: [{ field: 'general', message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
+    // MISSING TEST: Should verify description HTML validation works consistently with title
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:62]: **Missing test coverage**

The comment indicates missing test coverage for description HTML validation. This is important since there's a bug in the validation logic that could be caught with proper testing.

**Recommendation:** Add this test:
```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Overall Assessment

This PR adds comprehensive input validation to the Todo API, which is a significant security and reliability improvement. The implementation follows good practices with middleware-based validation, structured error responses, and extensive testing.

### ✅ **Strengths**

1. **Security Improvements**: HTML tag prevention mitigates XSS attacks
2. **Comprehensive Validation**: Covers all required fields with appropriate constraints
3. **Good Error Structure**: Structured error responses with field-specific messages
4. **Extensive Testing**: 15+ test cases covering various scenarios
5. **Documentation**: Well-documented validation rules in README
6. **Breaking Change Notice**: Properly documented API changes

### ⚠️ **Issues Found**

1. **Bug**: Inconsistent HTML validation regex between title and description
2. **Bug**: ID validation allows "0" which may be invalid
3. **Bug**: Missing error handling in POST route
4. **Missing Tests**: Gaps in test coverage for edge cases

### 📋 **Recommendations**

1. Fix the regex inconsistency in HTML validation
2. Clarify ID validation requirements (should "0" be valid?)
3. Add try-catch blocks around store operations
4. Complete test coverage for identified gaps
5. Consider adding input sanitization beyond validation

### 🎯 **Security Assessment**

The validation properly prevents:
- XSS through HTML tag blocking
- Type confusion attacks through type validation  
- Path traversal through ID format validation
- Input length attacks through size limits

### 📊 **Code Quality**

- Clean, readable middleware implementation
- Good separation of concerns
- Proper TypeScript typing
- Consistent error handling patterns

**Verdict**: This is a solid foundation that significantly improves the API's security posture. The bugs identified are fixable and don't compromise the overall value of the PR.
